storage/session_sql.go

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (C) 2024 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package storage
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"github.com/nuts-foundation/nuts-node/storage/log"
+	"gorm.io/gorm"
+	"gorm.io/gorm/schema"
+	"strings"
+	"sync"
+	"time"
+)
+
+var _ SessionDatabase = (*sqlSessionDatabase)(nil)
+
+var _ schema.Tabler = (*sessionStoreRecord)(nil)
+
+type sessionStoreRecord struct {
+	Store   string `gorm:"primaryKey"`
+	Key     string `gorm:"primaryKey;column:entry_key"`
+	Expires int
+	Value   string
+}
+
+func (s sessionStoreRecord) TableName() string {
+	return "session_store"
+}
+
+var sqlSessionPruneInterval = time.Minute
+
+func NewSQLSessionDatabase(db *gorm.DB) SessionDatabase {
+	result := sqlSessionDatabase{
+		db:       db,
+		routines: &sync.WaitGroup{},
+	}
+	result.ctx, result.cancel = context.WithCancel(context.Background())
+	result.startPruning()
+	return result
+}
+
+type sqlSessionDatabase struct {
+	db       *gorm.DB
+	routines *sync.WaitGroup
+	ctx      context.Context
+	cancel   context.CancelFunc
+}
+
+func (s sqlSessionDatabase) GetStore(ttl time.Duration, keys ...string) SessionStore {
+	return sqlSessionStore{
+		db:        s.db,
+		ttl:       ttl,
+		storeName: strings.Join(keys, "."),
+	}
+}
+
+func (s sqlSessionDatabase) close() {
+	s.cancel()
+	s.routines.Wait()
+}
+
+func (s sqlSessionDatabase) startPruning() {
+	s.routines.Add(1)
+	go func() {
+		defer s.routines.Done()
+		for {
+			select {
+			case <-time.After(sqlSessionPruneInterval):
+				if err := s.db.Exec("DELETE FROM session_store WHERE expires <= ?", time.Now().Unix()).Error; err != nil {
+					log.Logger().WithError(err).Warnf("SQL session store pruning failed")
+				}
+			case <-s.ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+type sqlSessionStore struct {
+	db        *gorm.DB
+	ttl       time.Duration
+	storeName string
+}
+
+func (s sqlSessionStore) Delete(key string) error {
```

Comment:
Describe behaviour when running multiple nodes with slightly different clocks.....

Reply:
I looked into using a database function for this, but there's no single approach (timestamp functions and/or datatypes) that works for all databases we support. Would it be better to just add a clock skew, e.g. 10 seconds (which should be plenty)?